File: android/app/src/main/java/com/phoneagent/core/product/ProductFingerprint.kt
package com.phoneagent.core.product

import java.security.MessageDigest

/**
 * Phone Agent - Step 2K Product Cryptographic Fingerprinting (Android)
 */
object ProductFingerprint {

    private fun sha256(input: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val hash = digest.digest(input.toByteArray(Charsets.UTF_8))
        return hash.joinToString("") { "%02x".format(it) }
    }

    fun computeProductFingerprint(product: ProductData): String {
        val sortedFeatures = product.keyFeatures.map { it.trim() }.filter { it.isNotEmpty() }.sorted().joinToString(";")
        val sortedBenefits = product.benefits.map { it.trim() }.filter { it.isNotEmpty() }.sorted().joinToString(";")
        val sortedSpecs = product.specifications.toSortedMap().entries.joinToString(";") { "${it.key}:${it.value.trim()}" }

        val canonical = listOf(
            product.productId?.trim().orEmpty(),
            product.source.trim().uppercase(),
            product.sourceUrl?.trim().orEmpty(),
            product.title.trim(),
            product.brand?.trim().orEmpty(),
            product.category?.trim().orEmpty(),
            product.price?.let { "%.2f".format(it) }.orEmpty(),
            product.currency.trim().uppercase(),
            product.priceQualifier?.trim().orEmpty(),
            product.availability?.trim().orEmpty(),
            product.rating?.let { "%.2f".format(it) }.orEmpty(),
            product.reviewCount?.toString().orEmpty(),
            product.description?.trim().orEmpty(),
            sortedFeatures,
            sortedBenefits,
            sortedSpecs,
            product.affiliateLink?.trim().orEmpty()
        ).joinToString("|")

        return "pfp_${sha256(canonical)}"
    }

    fun computeProductUrlFingerprint(url: String): String {
        if (url.isBlank()) return ""
        return "urlfp_${sha256(url.trim())}"
    }

    data class UrlValidationResult(val valid: Boolean, val reason: String? = null, val asin: String? = null)

    fun validateAmazonProductUrl(url: String): UrlValidationResult {
        val trimmed = url.trim()
        if (!trimmed.startsWith("https://") && !trimmed.startsWith("http://")) {
            return UrlValidationResult(false, "Unsupported URL scheme. Must use HTTPS.")
        }

        val allowedHosts = listOf(
            "amazon.com", "www.amazon.com", "amazon.in", "www.amazon.in",
            "amazon.co.uk", "www.amazon.co.uk", "amazon.de", "www.amazon.de",
            "amzn.to", "www.amzn.to", "amzn.in", "www.amzn.in"
        )

        val host = try {
            java.net.URI(trimmed).host?.lowercase() ?: return UrlValidationResult(false, "Invalid host syntax")
        } catch (e: Exception) {
            return UrlValidationResult(false, "Malformed URL")
        }

        val isAllowed = allowedHosts.any { host == it || host.endsWith(".$it") }
        if (!isAllowed) {
            return UrlValidationResult(false, "Host '$host' is not an allowed Amazon domain.")
        }

        val dpRegex = Regex("""/(?:dp|gp/product)/([A-Z0-9]{10})""", RegexOption.IGNORE_CASE)
        val match = dpRegex.find(trimmed)
        val asin = match?.groupValues?.get(1)?.uppercase()

        return UrlValidationResult(true, asin = asin)
    }
}

File: src/core/product/ProductFingerprint.ts
/**
 * Phone Agent - Step 2K Product Cryptographic Fingerprinting
 * Computes deterministic `pfp_<sha256>` fingerprints over canonicalized factual product data.
 * Validates Amazon URLs and computes `urlfp_<sha256>` for safe link extraction.
 */

import { sha256 } from '../content/ContentFingerprint';
import { ProductData } from './ProductData';

/**
 * Computes a deterministic cryptographic fingerprint for canonicalized product data.
 * Includes factual product attributes while excluding transient UI state.
 */
export function computeProductFingerprint(product: ProductData): string {
  // Sort features and specifications canonically
  const sortedFeatures = [...(product.keyFeatures || [])].map(f => f.trim()).filter(Boolean).sort();
  const sortedBenefits = [...(product.benefits || [])].map(b => b.trim()).filter(Boolean).sort();

  const specKeys = Object.keys(product.specifications || {}).sort();
  const sortedSpecs: Record<string, string> = {};
  for (const k of specKeys) {
    sortedSpecs[k] = (product.specifications?.[k] || '').trim();
  }

  // Canonical media asset IDs / checksums
  const mediaChecksums = (product.imageAssets || [])
    .map(a => a.sha256 || a.assetId)
    .sort();

  const canonicalObj = {
    productId: (product.productId || '').trim(),
    source: (product.source || '').trim().toUpperCase(),
    sourceUrl: (product.sourceUrl || '').trim(),
    title: (product.title || product.productName || '').trim(),
    brand: (product.brand || '').trim(),
    category: (product.category || '').trim(),
    price: product.price !== undefined ? Number(product.price.toFixed(2)) : null,
    currency: (product.currency || '').trim().toUpperCase(),
    priceQualifier: (product.priceQualifier || '').trim(),
    availability: (product.availability || '').trim(),
    rating: product.rating !== undefined ? Number(product.rating.toFixed(2)) : null,
    reviewCount: product.reviewCount !== undefined ? Math.floor(product.reviewCount) : null,
    description: (product.description || '').trim(),
    keyFeatures: sortedFeatures,
    benefits: sortedBenefits,
    specifications: sortedSpecs,
    mediaChecksums,
    affiliateLink: (product.affiliateLink || '').trim(),
  };

  return `pfp_${sha256(JSON.stringify(canonicalObj))}`;
}

/**
 * Computes a deterministic SHA-256 fingerprint for a canonicalized Amazon product URL.
 */
export function computeProductUrlFingerprint(url: string): string {
  if (!url || typeof url !== 'string') return '';
  const trimmed = url.trim();
  return `urlfp_${sha256(trimmed)}`;
}

/**
 * Validates Amazon product URL against strict safety invariants.
 * Allows amazon.com, amazon.in, amzn.to, etc.
 * Rejects malicious schemes, non-Amazon domains, or script injection.
 */
export function validateAmazonProductUrl(url: string): {
  valid: boolean;
  reason?: string;
  asin?: string;
  canonicalUrl?: string;
} {
  if (!url || typeof url !== 'string') {
    return { valid: false, reason: 'URL is empty or undefined.' };
  }

  const trimmed = url.trim();

  // Strict scheme check: https only (or http in testing mocks)
  if (!trimmed.startsWith('https://') && !trimmed.startsWith('http://')) {
    return { valid: false, reason: 'Unsupported URL scheme. Amazon URLs must use HTTPS.' };
  }

  try {
    const parsed = new URL(trimmed);
    const hostname = parsed.hostname.toLowerCase();

    const allowedHosts = [
      'amazon.com',
      'www.amazon.com',
      'amazon.in',
      'www.amazon.in',
      'amazon.co.uk',
      'www.amazon.co.uk',
      'amazon.de',
      'www.amazon.de',
      'amazon.ca',
      'www.amazon.ca',
      'amzn.to',
      'www.amzn.to',
      'amzn.in',
      'www.amzn.in',
      'a.co',
      'www.a.co',
    ];

    const isAllowedHost = allowedHosts.some(h => hostname === h || hostname.endsWith(`.${h}`));
    if (!isAllowedHost) {
      return { valid: false, reason: `Disallowed host: "${hostname}". Only verified Amazon domains are permitted.` };
    }

    // Extract ASIN if available (typically 10-char alphanumeric in /dp/B... or /gp/product/B...)
    let asin: string | undefined;
    const dpMatch = parsed.pathname.match(/\/(?:dp|gp\/product)\/([A-Z0-9]{10})/i);
    if (dpMatch && dpMatch[1]) {
      asin = dpMatch[1].toUpperCase();
    }

    return {
      valid: true,
      asin,
      canonicalUrl: trimmed,
    };
  } catch {
    return { valid: false, reason: 'Invalid URL syntax.' };
  }
}
